external/fv3fit/fv3fit/pytorch/cyclegan/cyclegan_trainer.py

```diff
@@ -0,0 +1,424 @@
+        return np.sqrt(self._sum_squared / self._count - self.mean ** 2)
+
+
+def get_r2(predicted, target) -> float:
+    """
+    Compute the R^2 statistic for the predicted and target data.
+    """
+    return 1.0 - np.var(predicted - target) / np.var(target)
+
+
+@dataclasses.dataclass
+class CycleGANTrainer:
+    """
+    A trainer for a CycleGAN model.
+
+    Attributes:
+        cycle_gan: the CycleGAN model to train
+        optimizer_generator: the optimizer for the generator
+        optimizer_discriminator: the optimizer for the discriminator
+        identity_loss: loss function used to make the generator which outputs
+            a given domain behave as an identity function when given data from
+            that domain as input
+        cycle_loss: loss function used on the difference between a round-trip
+            of the CycleGAN network and the original input
+        gan_loss: loss function used on output of the discriminator when
+            training the discriminator identify samples correctly or when training
+            the generator to fool the discriminator
+        batch_size: the number of samples to use in each batch when training
+        identity_weight: weight of the identity loss
+        cycle_weight: weight of the cycle loss
+        generator_weight: weight of the generator's gan loss
+        discriminator_weight: weight of the discriminator gan loss
+    """
+
+    # This class based loosely on
+    # https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/models/cycle_gan_model.py
+
+    # Copyright Facebook, BSD license
+    # https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/c99ce7c4e781712e0252c6127ad1a4e8021cc489/LICENSE
+
+    cycle_gan: CycleGAN
+    optimizer_generator: torch.optim.Optimizer
+    optimizer_discriminator: torch.optim.Optimizer
+    identity_loss: torch.nn.Module
+    cycle_loss: torch.nn.Module
+    gan_loss: torch.nn.Module
+    batch_size: int
+    identity_weight: float = 0.5
+    cycle_weight: float = 1.0
+    generator_weight: float = 1.0
+    discriminator_weight: float = 1.0
+
+    def __post_init__(self):
+        self.target_real: Optional[torch.autograd.Variable] = None
+        self.target_fake: Optional[torch.autograd.Variable] = None
+        self.fake_a_buffer = ReplayBuffer()
+        self.fake_b_buffer = ReplayBuffer()
+        self.generator_a_to_b = self.cycle_gan.generator_a_to_b
+        self.generator_b_to_a = self.cycle_gan.generator_b_to_a
+        self.discriminator_a = self.cycle_gan.discriminator_a
+        self.discriminator_b = self.cycle_gan.discriminator_b
+
+    def _init_targets(self, shape: Tuple[int, ...]):
+        self.target_real = torch.autograd.Variable(
+            torch.Tensor(shape).fill_(1.0).to(DEVICE), requires_grad=False
+        )
+        self.target_fake = torch.autograd.Variable(
+            torch.Tensor(shape).fill_(0.0).to(DEVICE), requires_grad=False
+        )
+
+    def evaluate_on_dataset(
```

Comment:
This function is a work in progress, it's unclear what metrics will be appropriate for automatically evaluating the quality of CycleGAN outputs given that there exists no paired output. I may entirely replace this function when I start using it on real data so I'd like to avoid iterating much at this stage.